### PR TITLE
Migrate to Kafka's new java client

### PR DESCRIPTION
**Pull request by asnare**
This pull request updates the Kafka sink to use the new Java client for Kafka. This is pretty straightforward, with the following caveats:

 - Although async in nature and API, some aspects of the new client are blocking.
    - The producer blocks on receiving cluster metadata before attempting to send.
    - The product can apparently block in some way when its internal buffer is full.
 - The new Java clients are not compatible with older server releases. Hence we use the 0.8 client instead of 0.9 for maximum server compatibility.
 - We now share a single producer amongst all threads, which may increase contention. (The Kafka folk believe that a single shared producer performs better than multiple producers, and it turns out that trying to use multiple producers results in exceptions as they collide try to register their JMX metrics.)
 - The new producer uses different property names to the old one. This means that upgrades from the old producer will need to rename some properties in their configuration.
 - We no longer fire-and-forget with messages to Kafka. This could result in a different performance profile.
 - We now LZ4-compress traffic that the producer sends to the brokers. This will increase the CPU consumption slightly, but we've noticed that often network bandwidth is at a higher premium than CPU.

### CODE REVIEW

File: src/main/java/io/divolte/server/kafka/KafkaFlusher.java [hunk 1]
```diff
@@ -83,55 +79,84 @@ public ProcessingDirective process(final Queue<AvroRecordBuffer> batch) {
...
+        // (In addition, we can be interrupted due to shutdown.)
+        final ImmutableList.Builder<ProducerRecord<DivolteIdentifier, AvroRecordBuffer>> remaining = ImmutableList.builder();
+        final int batchSize = batch.size();
+        for (int i = 0; i < batchSize; ++i) {
+            final Future<RecordMetadata> result = sendResults.get(i);
+            try {
+                final RecordMetadata metadata = result.get();
+                if (logger.isTraceEnabled()) {
+                    final ProducerRecord<DivolteIdentifier, AvroRecordBuffer> record = batch.get(i);
+                    logger.trace("Finished sending event (partyId={}) to Kafka: topic/partition/offset = {}/{}/{}",
```
Comment (friso) [inline src/main/java/io/divolte/server/kafka/KafkaFlusher.java]: We don't use the trace level elsewhere. Not sure if it is supported by all logging implementations.

**Comment by friso** [comment]
Looks good to me.

File: src/main/resources/reference.conf [hunk 2]
```diff
@@ -243,15 +243,15 @@ divolte {
...
       client.id = divolte.collector
       client.id = ${?DIVOLTE_KAFKA_CLIENT_ID}
 
-      request.required.acks = 0
-      message.send.max.retries = 5
-      retry.backoff.ms = 200
+      acks = 1
+      retries = 0
+      compression.type = lz4
+      max.in.flight.requests.per.connection = 1
```
Comment (cbsmith) [inline src/main/resources/reference.conf]: Need to set queue.enqueue.timeout.ms to 0.

Comment (cbsmith) [inline src/main/resources/reference.conf]: Also, I think you want to have a lot more in flight requests per connection as well as retries, given how the existing collector behaves in general.

Comment (cbsmith) [inline src/main/resources/reference.conf]: block.on.buffer.full needs to be set to false.

Sorry for the repost, accidentally deleted this comment during a moment of confusion.

Comment (friso) [inline src/main/resources/reference.conf]: These are defaults. Users can set them to anything else in their own deployments.

More than 1 in flight per connection implies the possibility of re-ordering in case of retry logic (in the case of multiple in-flights, it's undefined which error reaches our code first). We'll leave it to the user whether this is acceptable.

With `block.on.buffer.full` false, sending will result in a immediate exception when sending. The resulting exception is not retriable (`BufferExhaustedException` directly extends `KafkaException`, not `RetriableException`), meaning we'd immediately drop the event at that point instead of keeping it queued for a retry. The send buffer can be full for many reasons, some recoverable, some not. In either case, once messages made it to the send buffer, it's out of our control and in case of a unrecoverable failure on Kafka's side, we'll lose everything in the send buffer (but not everything in our own queue), without knowing whether it was sent or not. You can try to defeat this with retries being handled by the Kafka client automatically.

In short: the Kafka client attempts to do everything we need in terms of internal queueing and async usage, however it will in the end still block and throw exceptions nonetheless in the presence of distributed failures. In order to make sure that this does not interfere with other sinks (HDFS), we separate it out to a different thread with a queue in between.

Comment (cbsmith) [inline src/main/resources/reference.conf]: Divolte's code already implies the possibility of re-ordering in the case of retry logic, and this isn't configurable by the user. Setting to 1 in flight just (severely) reduces efficacy of compression as well as harms throughput if there is any network latency. Of course, having two levels of retry logic *will* cause some very confusing results no doubt.

Yes, you should drop the event instead of retrying, because there buffer is already full and retries have already been attempted. As per the logic in the ProcessingPool.enqueue(), when the buffer is full, you don't retry, and the only indication of a failure is a warning message with no context information about which event was dropped. So allowing a BufferExhaustedException to be trapped & logged by that code creates the same effect with less complexity and ambiguity.

Comment (asnare) [inline src/main/resources/reference.conf]: Compression is not impacted by the number of requests in flight. Note that `max.in.flight.requests.per.connection` controls the number of _requests_ that may be in-flight per connection with a broker, not the number of messages. The number of messages in the request will affect compression.

As things stand, these defaults yield:
 - Confidence messages have been delivered;
 - Less chance of reordering messages that relate to a given party in the face of temporary failures;
 - Protection against the Kafka producer blocking either due to metadata or the buffer being full.

As mentioned earlier, these are merely defaults. An administrator is free to prefer Kafka's retry mechanism, increase the concurrency, tweak `linger.ms`, adjust compression, and so on. There's nothing wrong with this, but the exact set of values are probably installation-specific. We're not telling you how to configure it for your purposes, but are providing a set of default values that should be safe and suitable for most users out-of-the-box.
